File: apps/api_v2/serializers/company.py
"""
Company serializers for API v2.
"""
from rest_framework import serializers

from apps.companies.models import Company


class CompanyV2Serializer(serializers.ModelSerializer):
    """
    Full company profile serializer.

    logo and signature are returned as absolute URLs when present.
    On update, use multipart/form-data to send image files.
    """

    logo_url = serializers.SerializerMethodField()
    signature_url = serializers.SerializerMethodField()

    class Meta:
        model = Company
        fields = [
            'id',
            'name',
            'email',
            'phone',
            'website',
            'address_line1',
            'address_line2',
            'city',
            'state',
            'postal_code',
            'country',
            'tax_id',
            'default_currency',
            'default_payment_terms',
            'default_tax_rate',
            'default_template',
            'default_notes',
            'accent_color',
            'invoice_prefix',
            'next_invoice_number',
            # Read-only image URLs (use dedicated upload endpoints to set images)
            'logo_url',
            'signature_url',
            # Late fee settings
            'late_fees_enabled',
            'late_fee_type',
            'late_fee_amount',
            'late_fee_grace_days',
            'late_fee_max_amount',
            # Stripe Connect (read-only status)
            'stripe_connect_onboarding_complete',
            'stripe_connect_charges_enabled',
            'stripe_connect_payouts_enabled',
            'created_at',
            'updated_at',
        ]
        read_only_fields = [
            'id',
            'next_invoice_number',
            'stripe_connect_onboarding_complete',
            'stripe_connect_charges_enabled',
            'stripe_connect_payouts_enabled',
            'logo_url',
            'signature_url',
            'created_at',
            'updated_at',
        ]

    def get_logo_url(self, obj):
        request = self.context.get('request')
        if obj.logo and request:
            return request.build_absolute_uri(obj.logo.url)
        return None

    def get_signature_url(self, obj):
        request = self.context.get('request')
        if obj.signature and request:
            return request.build_absolute_uri(obj.signature.url)
        return None

File: ios/InvoiceKits/InvoiceKits/InvoiceKits/Models/Company.swift
import SwiftData
import Foundation

// API response struct
struct CompanyResponse: Codable {
    let id: Int
    let name: String
    let email: String?
    let phone: String?
    let address: String?
    let city: String?
    let state: String?
    let zipCode: String?
    let country: String?
    let website: String?
    let taxId: String?
    let defaultPaymentTerms: String?
    let defaultNotes: String?
    let logoUrl: String?
    let signatureUrl: String?
}

// SwiftData cache model
@Model
final class CachedCompany {
    @Attribute(.unique) var serverId: Int
    var name: String
    var email: String?
    var phone: String?
    var address: String?
    var city: String?
    var state: String?
    var zipCode: String?
    var country: String?
    var logoUrl: String?

    init(from response: CompanyResponse) {
        self.serverId = response.id
        self.name = response.name
        self.email = response.email
        self.phone = response.phone
        self.address = response.address
        self.city = response.city
        self.state = response.state
        self.zipCode = response.zipCode
        self.country = response.country
        self.logoUrl = response.logoUrl
    }
}
